### DISCUSSION TITLE
Unable to increase storage size of kafka and zookeeper pods

### DISCUSSION THREAD

**Discussion by shreyasarani23** [strimzi/strimzi-kafka-operator#4956] [General]
In my storage class I have given allowvolumeexpansion as true
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: kafka-storage-class
provisioner: kubernetes.io/azure-disk
reclaimPolicy: Retain
allowVolumeExpansion: true
parameters:
  storageaccounttype: Standard_LRS
  kind: Managed
volumeBindingMode: WaitForFirstConsumer

And in my yaml file when I try to increase the size from 1Gi to 2Gi
it doesn't reflect in my pvc
 kubectl describe pvc data-my-kafka-cluster-kafka-0 -n kafka
Name:          data-my-kafka-cluster-kafka-0
Namespace:     kafka
StorageClass:  kafka-storage-class
Status:        Bound
Volume:        pvc-aa9f8636-5655-40d1-8a8a-51a59a2e76ef
Labels:        app.kubernetes.io/instance=my-kafka-cluster
               app.kubernetes.io/managed-by=strimzi-cluster-operator
               app.kubernetes.io/name=kafka
               app.kubernetes.io/part-of=strimzi-my-kafka-cluster
               strimzi.io/cluster=my-kafka-cluster
               strimzi.io/kind=Kafka
               strimzi.io/name=my-kafka-cluster-kafka
Annotations:   pv.kubernetes.io/bind-completed: yes
               strimzi.io/delete-claim: true
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      1Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       my-kafka-cluster-kafka-0
Events:        <none>
please help

And how do I restart my kafka pods?

**Comment by scholzj** [comment]
Can you please format the YAML snippets as code? It will make it much more readable and easier to see any issues or errors. In which YAML do you increase the size? You have to do it in the Kafka custom resource. You should probably also share the Cluster Operator logs to better understand what is going on.

**Reply by shreyasarani23** [reply]
I have attached my yaml file.
[kafka.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6471368/kafka.txt)
earlier storage was 1Gi , in the yaml file which I have attached I changed size 2Gi and deployed it. But is not reflecting in the pvc.
Below is the log of my cluster operator.
[clusterlog.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6471403/clusterlog.txt)

**Reply by scholzj** [reply]
Well, if you checked the log you can see that there is an error. So that is why it does not resize your storage. Plus the log is not complete, so hard to say why.  Is the Kafka custom resource really the resource you have in your Kubernetes cluster? Because it is not that you get with `kubectl get kafka -o yaml` and I do not see how it would cause this error.

**Reply by shreyasarani23** [reply]
Actually in my yaml I had jmx exporter and some metrics for kafka and zookeeper. Since I am not using it right now I removed it from yaml file and redeployed using kubectl apply -f kafka.yaml. And I can see some logs in the cluster operator related to changing the size from 1Gi to 2Gi. And when I describe my pvc I get the following error
 Warning  VolumeResizeFailed  109s (x17 over 7m18s)  volume_expand  error expanding volume "kafka/data-my-kafka-cluster-kafka-0" of plugin "kubernetes.io/azure-disk":azureDisk - disk resize is only supported on Unattached disk, current disk state: Attached, already attached to /subscriptions/b009a493-2b94-4f00-8e72-4556bff1c228/resourceGroups/MC_rg-infra-prototyping-1_teplay_centralindia/providers/Microsoft.Compute/virtualMachineScaleSets/aks-agentpool-34519435-vmss/virtualMachines/aks-agentpool-34519435-vmss_15.

And I have attached my cluster operator logs
[log.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6471687/log.txt)

**Reply by scholzj** [reply]
I'm not an Azure user. Normally, the resizing of the disk happens with mounted volumes. And later, if pod restart is needed to resize the filesystem, it is indicated in the status of the PVC.

**Reply by shreyasarani23** [reply]
how to restart the kafka pod?

**Reply by scholzj** [reply]
The operator log shows that it restarted. So I do not know what exactly does your environment need. In the docs you can find an annotation to trigger rolling update of the pods. You can also use `kubectl delete pod` I guess.

**Reply by shreyasarani23** [reply]
If I use kubectl delete pod my-kafka-cluster-kafka-0 then will it automatically deploy new one since it is a stateful set?
If so the existing pvc will be bound to the automatically create pod?

**Reply by scholzj** [reply]
That will always happen.

**Reply by shreyasarani23** [reply]
Do you recommend using kubectl rollout restart for statelful sets
ex:

kubectl rollout restart statefulset my-kafka-cluster-kafka -n kafka
kubectl rollout restart statefulset my-kafka-cluster-zookeeper -n kafka

**Reply by scholzj** [reply]
No, as I told you before, to do the rolling update, you should use the annotation from the docs.

**Reply by shreyasarani23** [reply]
ohk @scholzj

**Reply by shreyasarani23** [reply]
Hi @scholzj I did the following
1. I deleted my kafka cluster deployed. But my persistent volumes exists even if my cluster deleted. 
2. Since I have 3 kafka and zookeeper pods I created 6 PVC's  linking the PVCs to the appropriate PV and I changed storage to 2Gi 
3. Next I edited my PV's by removing claimref section and changed storage:2Gi for all 6 PV's
4. Deployed my kafka cluster
5. when I do kubectl get pv I get this
![pv-kafka](https://user-images.githubusercontent.com/35260380/118128952-2ef62f00-b419-11eb-9988-7b9ff1a9bbff.JPG)
6. when i do kubectl get pvc I get this
![pvc-kafka](https://user-images.githubusercontent.com/35260380/118129008-4208ff00-b419-11eb-8768-c7e585110f48.JPG)
7. But now when I enter into pod by executing kubectl exec -it my-kafka-cluster-kafka-0  -n kafka -- /bin/bash and execute df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay         124G   19G  106G  15% /
tmpfs            64M     0   64M   0% /dev
tmpfs           3.4G     0  3.4G   0% /sys/fs/cgroup
tmpfs           3.4G   52K  3.4G   1% /tmp
shm              64M     0   64M   0% /dev/shm
/dev/sda1       124G   19G  106G  15% /etc/hosts
tmpfs           3.4G     0  3.4G   0% /var/opt/kafka
tmpfs           3.4G   12K  3.4G   1% /opt/kafka/client-ca-certs
tmpfs           3.4G   48K  3.4G   1% /opt/kafka/broker-certs
tmpfs           3.4G   12K  3.4G   1% /opt/kafka/cluster-ca-certs
/dev/sdd        976M  3.2M  957M   1% /var/lib/kafka/data

I can see that /dev/sdd is still 976M instead of 2Gi.
8. And also in azure portal it has not reflected the size of disk. Still it is showing 1Gi
![azure](https://user-images.githubusercontent.com/35260380/118129767-366a0800-b41a-11eb-9045-79dd6b4ada02.JPG)

I don't know why this happened. please tell me if there is any mistakes.
Please help me to resolve this issue

**Reply by scholzj** [reply]
I'm sorry ... I described the standard process of increasing the volume size on Kubernetes with Strimzi above, it is also in the docs. But I do not know anything about your infrastructure or cluster configuration. And your description does not really make it clear what exactly have you done and when.

**Reply by shreyasarani23** [reply]
I have followed this link https://strimzi.io/docs/operators/in-development/using.html#proc-resizing-persistent-volumes-str
but it didn't work for me it showed me the error that it cannot resize when the disk in attached state. So I did the above steps .
Can you give me the link where they have explained increasing volume size in kubernetes .

**Reply by shreyasarani23** [reply]
I was able to increase the size of my PV by following this link https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/. Thanks @scholzj for the help.

**Reply by scholzj** [reply]
Well, the Kubernetes link is exactly the same what the operator does. So I'm glad it helped but not sure it explains why did it not worked for you before.
